## Students Information

Please enter the names and IDs of the two students below:

1. **Name**: [Enter Student 1 Name Here]  
   **ID**: `9XXXXXX` 

2. **Name**: [Enter Student 2 Name Here]  
   **ID**: `9XXXXXX` 


## Students Instructions

This is your first graded lab assignment, as you put the work you have studied in the lectures in action, please take this opportunity to enhance your understanding of the concepts and hone your skills. As you work on your assignment, please keep the following instructions in mind:

- Clearly state your personal information where indicated.
- Be ready with your work before the time of the next discussion slot in the schedule.
- Plagiarism will be met with penalties, refrain from copying any answers to make the most out of the assignment. If any signs of plagiarism are detected, actions will be taken.
- It is acceptable to share the workload of the assignment bearing the discussion in mind.
- Feel free to [reach out](mailto:cmpsy27@gmail.com) if there were any ambiguities or post on the classroom.

## Submission Instructions

To ensure a smooth evaluation process, please follow these steps for submitting your work:

1. **Prepare Your Submission:** Alongside your main notebook, include any additional files that are necessary for running the notebook successfully. This might include data files, images, or supplementary scripts.

2. **Rename Your Files:** Before submission, please rename your notebook to reflect the IDs of the two students working on this project. The format should be `ID1_ID2`, where `ID1` and `ID2` are the student IDs. For example, if the student IDs are `9123456` and `9876543`, then your notebook should be named `9123456_9876543.ipynb`.

3. **Check for Completeness:** Ensure that all required tasks are completed and that the notebook runs from start to finish without errors. This step is crucial for a smooth evaluation.

4. **Submit Your Work:** Once everything is in order, submit your notebook and any additional files via the designated submission link on Google Classroom **(code: 2yj6e24)**. Make sure you meet the submission deadline to avoid any late penalties.
5. Please, note that the same student should submit the assignments for the pair throughout the semester.

By following these instructions carefully, you help us in evaluating your work efficiently and fairly **and any failure to adhere to these guidelines can affect your grades**. If you encounter any difficulties or have questions about the submission process, please reach out as soon as possible.

We look forward to seeing your completed projects and wish you the best of luck!





## Installation Instructions

In this lab assignment, we require additional Python libraries for scientific mathematics, particularly in the context of machine learning (ML) and satellite image analysis. To fulfill these requirements, we need to install Scikit-learn and Scikit-image. 
1. Install Scikit-learn  
Scikit-learn (Sklearn) is a powerful Python library for ML tasks, offering various algorithms for classification, regression, clustering, and model evaluation. It is extensively used for analyzing satellite imagery, enabling tasks such as land cover classification and environmental parameter prediction. On the other hand, Scikit-image (Skimage) provides comprehensive tools for image processing and computer vision, facilitating tasks such as image preprocessing, feature extraction, and segmentation. These libraries are essential for extracting valuable insights from satellite images and conducting advanced analysis in scientific computing and research domains.
```bash
pip install scikit-learn scikit-image
```


> **Note:** You are allowed to install any other necessary libraries you deem useful for solving the lab. Please ensure that any additional libraries are compatible with the project requirements and are properly documented in your submission.


## Maximum Likelihood Estimator (MLE) Classifier
The Maximum Likelihood Estimator (MLE) is a fundamental statistical approach used to infer the parameters of a given distribution that are most likely to result in the observed data. In the context of image classification, MLE helps to quantify the probability of observing the data within each predefined class based on their distinct statistical properties. This method is highly effective for classifying images into categories by comparing the likelihoods of the data under different model parameters, enabling the most probable class assignment.

1. **Calculate Class Priors**: Estimate the probability of each class based on the training dataset. This is expressed as:
   $$
   P(C_k) = \frac{N_k}{N}
   $$
   where \(N_k\) is the number of samples of class \(k\) and \(N\) is the total number of samples.

2. **Estimate Class-specific Parameters**: For each class, estimate parameters such as the mean \(\mu_k\) and covariance \(\Sigma_k\) of features that describe the distribution of the data:
   $$
   \mu_k = \frac{1}{N_k} \sum_{x \in C_k} x
   $$
   $$
   \Sigma_k = \frac{1}{N_k} \sum_{x \in C_k} (x - \mu_k)(x - \mu_k)^T
   $$

3. **Compute Likelihoods**: For a given test instance \(x\), compute the likelihood of that instance belonging to each class using the estimated parameters:
   $$
   p(x | C_k) = \frac{1}{(2\pi)^{d/2} |\Sigma_k|^{1/2}} \exp\left(-\frac{1}{2} (x - \mu_k)^T \Sigma_k^{-1} (x - \mu_k)\right)
   $$

4. **Classify Based on Maximum Likelihood**: Assign the class label to each test instance based on the highest likelihood, which can be calculated as:
   $$
   \hat{y} = \arg\max_{k} P(C_k) \cdot p(x | C_k)
   $$

The Naive Bayes classifier is perhaps the most well-known application of the Maximum Likelihood Estimator principle in classification tasks. It assumes that the features in each class are independent, simplifying the computation of likelihoods. While Naive Bayes is popular for its simplicity and efficiency, it is not the only technique that leverages the MLE approach. Other classical alternatives include Logistic Regression, which applies MLE to estimate the parameters that best predict categorical outcomes, and Gaussian Mixture Models, which use MLE to estimate the parameters of multiple Gaussian distributions within the data. Students are encouraged to explore these models to gain a deeper understanding of statistical estimation techniques.


## Req- Image Classification for EuroSATallBands
Image classification is a key challenge in satellite imaging and remote sensing. As discussed in the lecture, this task is typically conducted on a pixel-wise basis because a single image can contain multiple textural elements of different celestial features. However, for this specific assignment, we will focus on identifying the dominant phenomena in the image as the basis for classification.

- **Load the Images**: Load the images of the EuroSAT dataset that belong to the **residential**, **river**, and **forest** classes.

- **Split the Dataset**: Split the dataset such that 10% of each class is used as testing data, and the remainder is used for training your classifier. Use the indices provided by `np.random.choice` with seed set to `27`. **Code is provided do not change it**.

- **Feature Extraction**: Extract suitable features from the images that you think might be relevant in distinguishing each class from the others. Keep in mind the curse of dimensionality when selecting features.

- **Implement a Maximum Likelihood Estimator (MLE)**: Implement a Maximum Likelihood Estimator (MLE) based on your training data. 
- **Report Accuracy and Average F1 Score**: After testing your classifier on the test set, report the **Accuracy** and **Average F1 Score** of your model.


In [9]:
# Add your libraries here
import numpy as np
from skimage import io
import os




In [10]:
# DO NOT CHANGE THIS CELL
## Training set indices.
np.random.seed(27)  # Set random seed for reproducibility

# Randomly select indices for the test sets for each class
residential_test_indices = np.random.choice(np.arange(3000), size=300, replace=False)
forest_test_indices = np.random.choice(np.arange(3000), size=300, replace=False)
river_test_indices = np.random.choice(np.arange(2500), size=250, replace=False)


In [11]:
# Follow the steps 
print(residential_test_indices )


[1748 1976 2825 1611  404  502  843 2977 1657  168 2972 2989   88  110
 1853 1121 2640 1075 2398 1705  809 1231  466 2068 1545  713 1780  493
 1884 2118 2020 1097 1601 1610 1456 1394 2472  602 1617 2809  706 2641
  592  108  351 1752  401 1233 2269  393 1720   64 2883 2504  828 2837
  293   75 2584 1994  289 1903  615 1005 1700 2770  389  472 2686   43
 1528  232  500 2819 1852 1531 2084 1498 1722 2387  877  601 1640  583
 2209 1536 2939 1411 1057 2391 2918   45  253 1189 2488 2991  366 2214
 1638  710  529  211 2028  418 2895 2548 2364  195  552 1192 1403 1653
  468 2956 1949 2901 2854 2047  632 1261 2790  188 1726 2151 2396 1204
 1494 2145  489 2846 2355 2104 1627 2228 2194 1733  691 2434  823 2799
 2291 1584 2802  819  674   95  985 1093  642 1644 2539 1408 2656 2751
 2927 2574  643  191 1680 2486 2630 1489 1918 1119 2162  955 1905 1191
 1571  785 1407  395 2832  203 2362  438 2468 2378 2661 1220 1459 1254
  735 1225 1386  378 1018 1523 2739  867  423  202 1755 1108 2764  420
 2881 

# Load the Images

In [12]:
dataset_folder = "D:/Astudy/fourth year/second term/satellite imagery/dataset/EuroSATallBands/EuroSATallBands/ds/images/remote_sensing/otherDatasets/sentinel_2/tif" 
classes = ["residential", "river", "forest"]
images_by_class = {}

for class_name in classes:
  
    class_folder = os.path.join(dataset_folder, class_name)
    if os.path.exists(class_folder):
        
        images_by_class[class_name] = []
        for image_file in os.listdir(class_folder):
            
            image_path = os.path.join(class_folder, image_file)
            image = io.imread(image_path)
            image = image.reshape(image.shape[0] * image.shape[1], image.shape[2])
            images_by_class[class_name].append(image)



residential_images = np.array(images_by_class["residential"])

river_images = np.array(images_by_class["river"])

forest_images = np.array(images_by_class["forest"])


print(residential_images)



[[[1482 1088 1051 ... 2077 1236 2429]
  [1482 1088 1051 ... 2077 1236 2429]
  [1480 1145 1018 ... 1801 1084 2184]
  ...
  [1392 1138 1068 ... 1676 1163 2475]
  [1394 1240 1080 ... 1694 1198 2476]
  [1397 1262 1295 ... 1698 1192 2595]]

 [[1370 1079  956 ... 1641 1132 2266]
  [1370 1079  956 ... 1641 1132 2266]
  [1376 1013  901 ... 1630 1147 2203]
  ...
  [1411 1073  971 ... 1729 1124 2435]
  [1410 1108  880 ... 1704 1111 2343]
  [1410 1073  881 ... 1739 1166 2263]]

 [[1393 1235 1168 ... 2649 1682 3508]
  [1393 1235 1168 ... 2649 1682 3508]
  [1394 1313 1371 ... 2731 1839 3342]
  ...
  [1355 1163 1151 ... 2532 1445 3890]
  [1351 1204 1190 ... 2502 1363 3966]
  [1349 1097 1076 ... 2445 1274 3954]]

 ...

 [[1601 1519 1386 ... 2326 1844 2167]
  [1601 1519 1386 ... 2326 1844 2167]
  [1591 1601 1568 ... 2472 1966 2265]
  ...
  [1646 1520 1493 ... 2630 2029 2693]
  [1650 1546 1442 ... 2599 2012 2599]
  [1652 1412 1250 ... 2548 1966 2530]]

 [[1162  848  748 ... 1531  845 3809]
  [1162  848

 # Split the Dataset

In [13]:
print(len(residential_images))
print(len(river_images))
print(len(forest_images))



# print(residential_test_indices)
test_residual = residential_images[residential_test_indices]
test_river = river_images[river_test_indices]
test_forest =  forest_images[forest_test_indices]

# print(test_residual)
# Generate the training indices by excluding the test indices
residential_training_indices = np.setdiff1d(np.arange(3000), residential_test_indices)
forest_training_indices = np.setdiff1d(np.arange(3000), forest_test_indices)
river_training_indices = np.setdiff1d(np.arange(2500), river_test_indices)

train_residual = residential_images[residential_training_indices]
train_river = river_images[river_training_indices]
train_forest =forest_images[forest_training_indices]

print(train_residual.shape)
print(train_river.shape)
print(train_forest.shape)


3000
2500
3000
(2700, 4096, 13)
(2250, 4096, 13)
(2700, 4096, 13)


In [14]:
print(train_residual.shape[0])
print(train_river.shape)
print(train_forest.shape)


2700
(2250, 4096, 13)
(2700, 4096, 13)


# Feature Extraction

# Implement a Maximum Likelihood Estimator (MLE)

In [15]:
def get_Class_Priors(train_residual,train_forest,train_river):
  total_len = train_forest.shape[0] + train_residual.shape[0] + train_river.shape[0]
  p_residual =  train_residual.shape[0]/total_len
  p_river =  train_river.shape[0]/total_len
  p_forest = train_forest.shape[0]/total_len

  return p_residual,p_river,p_forest

p_residual,p_river,p_forest = get_Class_Priors(train_residual,train_forest,train_river)
print("p_residual: ",p_residual)
print("p_river: ",p_river)
print("p_forest: ",p_forest)

print("sum prob: ",p_forest+p_residual+p_river)

p_residual:  0.35294117647058826
p_river:  0.29411764705882354
p_forest:  0.35294117647058826
sum prob:  1.0


In [16]:
def Estimate_Class_specific_Parameters(class_array):
    images_mean = np.zeros(class_array[0].shape[1])
    for i in range(class_array.shape[0]):
        image = class_array[i]
        image_mean =  np.mean(image, axis=0)
       
        images_mean += image_mean
    # print(images_mean)
    class_mean = images_mean/ class_array.shape[0]
    # new_image = []
    new_image = np.zeros((13, 13))
    for i in range (class_array.shape[0]):
        x_minus_mean = []
        image = class_array[i]
        for pixel in image: 
            new_val = np.array(pixel-class_mean)
            new_val_mul = np.outer(new_val ,new_val) 
            # print(new_val_mul)
            # print(new_val_mul.shape )
            # break
            # new_image.append( new_val_mul )
            new_image += new_val_mul
        # new_image.append(x_minus_mean)    
        # print("x_minus mean",x_minus_mean[0])
        # print("x",pixel)
        # break
    # sum = np.sum(new_image,axis=1)  
    class_covariance =   new_image / class_array.shape[1]
    # print(new_image)
    # print(class_mean)
    # print(class_array.shape[0])
    # print(class_array[0].shape[1])
    return class_mean  ,  class_covariance     

forest_mean ,forest_covariance = Estimate_Class_specific_Parameters(train_forest)
# river_mean = Estimate_Class_specific_Parameters(train_river)
# residual_mean = Estimate_Class_specific_Parameters(train_residual)

print("forest_mean",forest_mean)
print("forest_covariance",forest_covariance)
# print("river_mean",river_mean)
# print("residual_mean",residual_mean)
    

forest_mean [1109.91998806  811.50463216  687.73663131  415.79315457  748.9670667
 2100.11004078 2688.02510932 2634.13981834  826.21561026    9.93443043
 1312.54340368  535.46771991 3004.78895924]
forest_covariance [[ 3.23623931e+07  2.42506742e+07  1.04553375e+07  1.31021992e+07
  -9.48200102e+04 -4.43495981e+07 -5.93573854e+07 -6.05855647e+07
  -3.86147251e+07 -2.23298155e+03 -1.25111736e+07 -1.51247093e+06
  -6.02606564e+07]
 [ 2.42506742e+07  2.05906884e+07  1.35083695e+07  1.47678928e+07
   8.98728362e+06 -1.89645741e+07 -2.94793129e+07 -2.51558495e+07
  -2.05882354e+07  8.97556940e+04  2.58978464e+06  5.90651394e+06
  -2.86754253e+07]
 [ 1.04553375e+07  1.35083695e+07  2.73572030e+07  1.86126497e+07
   4.55977928e+07  6.52344945e+07  6.44948303e+07  8.12760791e+07
   4.25764231e+07  5.35392781e+05  5.09244568e+07  2.91436617e+07
   7.30709941e+07]
 [ 1.31021992e+07  1.47678928e+07  1.86126497e+07  1.96005946e+07
   2.75735092e+07  1.31129527e+07  3.76662397e+06  1.28729971e+07
  

In [18]:
# def Compute_Likelihoods(mean,covariance):
        

In [ ]:
# def Classify_Based_on_Maximum_Likelihood():

## Submission Instructions

To ensure a smooth evaluation process, please follow these steps for submitting your work:

1. **Prepare Your Submission:** Alongside your main notebook, include any additional files that are necessary for running the notebook successfully. This might include data files, images, or supplementary scripts.

2. **Rename Your Files:** Before submission, please rename your notebook to reflect the IDs of the two students working on this project. The format should be `ID1_ID2`, where `ID1` and `ID2` are the student IDs. For example, if the student IDs are `9123456` and `9876543`, then your notebook should be named `9123456_9876543.ipynb`.

3. **Check for Completeness:** Ensure that all required tasks are completed and that the notebook runs from start to finish without errors. This step is crucial for a smooth evaluation.

4. **Submit Your Work:** Once everything is in order, submit your notebook and any additional files via the designated submission link on Google Classroom **(code: 2yj6e24)**. Make sure you meet the submission deadline to avoid any late penalties.
5. Please, note that the same student should submit the assignments for the pair throughout the semester.

By following these instructions carefully, you help us in evaluating your work efficiently and fairly **and any failure to adhere to these guidelines can affect your grades**. If you encounter any difficulties or have questions about the submission process, please reach out as soon as possible.

We look forward to seeing your completed projects and wish you the best of luck!


# Report Accuracy and Average F1 Score

### Grading Rubric (Total: 10 Marks)

The lab is graded based on the following criteria:

1. **Data Loading and Preparation (2 Marks)**
   - Correctly loads images for the residential, river, and forest classes. (0.5 Marks)
   - Accurately splits the dataset into training and testing subsets and clearly shows this split. (1.5 Marks)

2. **Feature Extraction (2 Marks)**
   - Implements feature extraction appropriately, considering the curse of dimensionality. (1 Mark)
   - Extracts and justifies the selection of features relevant to distinguishing the classes. (1 Mark)

3. **Implementation of MLE Classifier (3 Marks)**
   - Correctly calculates and clearly shows class priors and class-specific parameters. (1 Mark)
   - Accurately computes likelihoods using the likelihood equation (probability density function) and classifies based on maximum likelihood. Must clearly show these calculations and explain the choice of likelihood equation. (2 Marks)

4. **Model Evaluation and Understanding (3 Marks)**
   - Shows **confusion matrix** and correctly calculates and clearly shows the calculations for Accuracy and Average F1 Score. (1 Mark)
   - **Comparison amongst your peers.** Compares the model's performance against those of peers to identify strengths and areas for improvement. (2 Marks)

Each section of the lab will be evaluated on completeness, and correctness in approach and analysis. Part of the rubric also includes the student's ability to explain and justify their choices and results.
